### PR TITLE
Allow deleteing of instructor-lecture associtaions

### PR DESCRIPTION
**Pull request by drouba**

### CODE REVIEW

File: public/instructors_list.php [hunk 1]
```diff
@@ -56,6 +56,11 @@
     header("refresh: 1");
 }
 
+if (isset($_POST['delete_association']) && LectureInstructor::find_by(array('lecture_id' => $_POST['lecture_id'], 'user_id' => $_POST['instructor_id'])) != null) {
+    LectureInstructor::find_by(array('lecture_id' => $_POST['lecture_id'], 'user_id' => $_POST['instructor_id']))->deleteWhere('user_id', 'lecture_id');
+    header("refresh: 1");
```
Comment (jamestjw) [inline public/instructors_list.php:61]: I approve of the idea of implementing a delete_where (but without fetching the records first)

File: public/instructors_list.php [hunk 2]
```diff
@@ -120,14 +125,18 @@
...
             <tbody>
         <?php foreach ($course_assignment as $row) {?>
             <tr>
-                <td><?php echo $row->lecture->id; ?></td>
+                <td><?php echo $row->lecture->course->course_name." ".$row->lecture->lecture_code; ?></td>
                 <td><?php echo $row->user->get_full_name(); ?></td>
                 <td><?php echo escape($row->created_at);  ?> </td>
+                <td><form method="post"><input type="checkbox" name='lecture_id' value="<?php echo$row->lecture->id; ?>"><input type="submit" name="delete_association" value="delete">
+                    <input type="hidden" name='instructor_id' value="<?php echo $row->user->id;?>">
+                    </form></td>
```
Comment (jamestjw) [inline public/instructors_list.php:139]: why do we need a checkbox?

**Review by jamestjw — CHANGES_REQUESTED** [review]

File: public/instructors_list.php (hunk 2 above)
Comment (drouba) [reply public/instructors_list.php:139]: good point, I'll change that

File: public/instructors_list.php (hunk 1 above)
Comment (drouba) [reply public/instructors_list.php:61]: not sure how to do this :S 
It works the way it is now

Comment (jamestjw) [reply public/instructors_list.php:61]: where is the function? can't find it

Comment (drouba) [reply public/instructors_list.php:61]: it should be in records just after the delete function

Comment (drouba) [reply public/instructors_list.php:61]: huh it wasnt added again weird

Comment (drouba) [reply public/instructors_list.php:61]: added it

File: public/instructors_list.php (hunk 2 above)
Comment (drouba) [reply public/instructors_list.php]: done

File: modules/models/record.php [hunk 3]
```diff
@@ -123,6 +123,36 @@ public function delete($pk="id") // Deletes by assuming $id is always the pk. No
...
+
+        $sql = sprintf(
+            "DELETE FROM %s WHERE %s = ? AND %s = ?",
+            get_called_class()::$table_name,
+            $pk, $pk2
+        );
+
+        execute_sql_query($sql, array($pk => $this->$pk, $pk2=>$this->$pk2));
+        return;
+    }
```
Comment (jamestjw) [inline modules/models/record.php:154]: this return is kind of redundant 😅

File: modules/models/record.php [hunk 4]
```diff
@@ -123,6 +123,36 @@ public function delete($pk="id") // Deletes by assuming $id is always the pk. No
         return;
     }
 
+    public function deleteWhere($pk, $pk2) // Deletes from tables where pk is compound key
+    {
+        // Check to see if we are deleting a record that has not been saved yet.
+        if ($this->is_new_record) {
+            throw new ErrorException("Cannot delete an object that is not yet in a table. Please set the object to null instead.");
+        }
+
```
Comment (jamestjw) [inline modules/models/record.php:132]: i don't really like this function, I expected it to have a function signature similar to `where`, but im ok with overlooking this

File: modules/models/record.php [hunk 5]
```diff
@@ -123,6 +123,36 @@ public function delete($pk="id") // Deletes by assuming $id is always the pk. No
...
+        // Ensure that the pk exists
+        if (!in_array($pk, $attrs) || !in_array($pk2, $attrs)) {
+            throw new ErrorException("Record trying to delete using a primary key that does not exist: ".$pk);
+        }
+
+        $result = get_called_class()::where(array($pk=>$this->$pk, $pk2=>$this->$pk2));
+        // Make sure that only one record with the supplied key exists in the table
+        if (count($result) > 1) {
+            throw new ErrorException("Primary key specified is not valid as it returns more than one record. PK: ".$pk);
```
Comment (jamestjw) [inline modules/models/record.php:143]: this doesn't cover the second pk btw

**Review by jamestjw — APPROVED** [review]

**Pull request merged**